crates/runtime/src/dataconnector/s3.rs

```diff
@@ -208,31 +212,35 @@ impl DataConnectorFactory for S3Factory {
-                        return Err(Box::new(Error::InvalidAuthParameterCombination {
-                            parameter: "s3_key".to_string(),
-                            auth: "key".to_string(),
-                        })
-                            as Box<dyn std::error::Error + Send + Sync>);
-                    }
-                    if matches!(params.parameters.get("secret"), ParamLookup::Present(_)) {
-                        // The 's3_secret' parameter cannot be set unless the `s3_auth` parameter is set to 'key'.
-                        return Err(Box::new(Error::InvalidAuthParameterCombination {
-                            parameter: "s3_secret".to_string(),
-                            auth: "key".to_string(),
-                        })
-                            as Box<dyn std::error::Error + Send + Sync>);
+                    // These parameters cannot be set unless the `s3_auth` parameter is set to 'key'.
+                    for param in ["key", "secret", "session_token"] {
+                        if matches!(params.parameters.get(param), ParamLookup::Present(_)) {
+                            return Err(Box::new(Error::InvalidAuthParameterCombination {
+                                parameter: format!("{PREFIX}_{param}"),
+                                auth: "key".to_string(),
+                            })
+                                as Box<dyn std::error::Error + Send + Sync>);
+                        }
                     }
                 }
                 Some("key") => {
-                    if matches!(params.parameters.get("key"), ParamLookup::Absent(_)) {
-                        return Err(Box::new(Error::NoAccessKey)
-                            as Box<dyn std::error::Error + Send + Sync>);
+                    for (param, e) in [
+                        ("key", Error::NoAccessKey),
+                        ("secret", Error::NoAccessSecret),
+                    ] {
+                        if matches!(params.parameters.get(param), ParamLookup::Absent(_)) {
+                            return Err(Box::new(e) as Box<dyn std::error::Error + Send + Sync>);
+                        }
                     }
-                    if matches!(params.parameters.get("secret"), ParamLookup::Absent(_)) {
-                        return Err(Box::new(Error::NoAccessSecret)
-                            as Box<dyn std::error::Error + Send + Sync>);
+
+                    // Session token is optional; log if present
+                    if let ParamLookup::Present(secret) = params.parameters.get("session_token") {
+                        tracing::info!(
+                            "Using temporary credentials with session token for S3 auth."
+                        );
+
+                        // Set environment variable because session token cannot be used in URL fragment
+                        std::env::set_var("AWS_SESSION_TOKEN", secret.expose_secret());
```

Comment:
This shouldn't set an env var and instead should set it on the fragment, and set it on the object store instance from the fragment in `object_store_registry.rs`